Sources/PackagePlugin/Context.swift

```diff
@@ -6,14 +6,64 @@
 
  See http://swift.org/LICENSE.txt for license information
  See http://swift.org/CONTRIBUTORS.txt for Swift project authors
- */
+*/
+
+/// Provides information about the package for which the plugin is invoked,
+/// as well as contextual information based on the plugin's stated intent
+/// and requirements.
+public struct PluginContext {
+    /// Information about the package to which the plugin is being applied.
+    public let package: Package
+
+    /// The path of a writable directory into which the plugin or the build
+    /// commands it constructs can write anything it wants. This could include
+    /// any generated source files that should be processed further, and it
+    /// could include any caches used by the build tool or the plugin itself.
+    /// The plugin is in complete control of what is written under this di-
+    /// rectory, and the contents are preserved between builds.
+    ///
+    /// A plugin would usually create a separate subdirectory of this directory
+    /// for each command it creates, and the command would be configured to
+    /// write its outputs to that directory. The plugin may also create other
+    /// directories for cache files and other file system content that either
+    /// it or the command will need.
```

Comment:
Random question: what's the state of creating temporary directories in Swift? We have some code for this in `TSCBasic` and this being mentioned here made me wonder where we should offer some API to make that easier for plugin authors.

Reply:
That's a good point — you mean something like NSTemporaryDirectory(), or something inside this directory here?  I think the idea would be that the plugin should get the directories passed to it for both cache and temporary location.

Long-term I'd prefer if the plugin didn't access the file system at all, but rather used API provided by the plugin host.  That would also make it easier to make things more distributable.  But the need to shell out to local tools might actually spoil that.